src/core/ext/transport/chttp2/server/chttp2_server.cc

```diff
@@ -774,18 +790,38 @@ grpc_error_handle Chttp2ServerListener::CreateWithAcceptor(
   // TODO(yangg) channelz
   TcpServerFdHandler** arg_val = args.GetPointer<TcpServerFdHandler*>(name);
   *arg_val = grpc_tcp_server_create_fd_handler(listener->tcp_server_);
-  server->AddListener(OrphanablePtr<Server::ListenerInterface>(listener));
+  server->AddListener(OrphanablePtr<ListenerInterface>(listener));
   return absl::OkStatus();
 }
 
+absl::StatusOr<ListenerInterface*>
+Chttp2ServerListener::CreateForPassiveListener(
+    Server* server, const ChannelArgs& args,
+    Chttp2ServerArgsModifier args_modifier) {
+  Chttp2ServerListener* listener =
+      new Chttp2ServerListener(server, args, args_modifier, nullptr);
+  grpc_error_handle error = grpc_tcp_server_create(
```

Comment:
Oh, interesting!  Neither `ListenerInterface` nor `Chttp2ServerListener` are derived from any of the ref-counted types, so I assumed that they were not already ref-counted.  I didn't realize that they implemented their own ref-counting based on the iomgr TCP server.

I think that changes the decision we made in https://github.com/grpc/grpc/pull/35957#discussion_r1531088144, where we decided to add a listener lock to the core server based partially on the assumption that the listener was not already ref-counted, which now turns out not to be true.  Given that, I think I want to go the other way: instead of adding a lock in the server, let's have `PassiveListenerImpl` hold a ref to the listener.

With respect to the question here, note that this use of ref-counting based on the iomgr TCP server is going to have to be replaced with a separate ref count when we migrate to EvenEngine anyway, since the EE listener object is not ref-counted.  I am totally willing to accept a temporary small memory increase to have a separate ref-count between now and when we finish the EE migration.

Given all of that, here's what I suggest:
- Change `ListenerInterface` to inherit from `InternallyRefCounted<>` instead of from `Orphanable`.
- In `Chttp2ServerListener`, remove the custom `Ref()` and `Unref()` methods.  Instead, whereever we call `grpc_tcp_server_create()`, take a ref to `Chttp2ServerListener`, and release that ref in `TcpServerShutdownComplete()`.
- Revert the change to add the listener lock in the core server.
- Change `PassiveListenerImpl` to hold a `RefCountedPtr<ListenerInterface>`.  I think this will also require moving `PassiveListenerImpl` from the C++ layer to the C-core layer, because that will allow the code here to set this field in `PassiveListenerImpl` without exposing `RefCountedPtr<>` in the C-core API.  (See [the code sketch I provided in my original review pass](https://github.com/grpc/grpc/pull/35957#discussion_r1525606284) for what this might look like.)

I think this should eliminate the need to create a TCP server for the passive listener.

Reply:
I like moving away from that design, removing the lock saves a big lock acquisition ordering headache. 

The public C++ API relies on `ListenerInterface`, in `include/grpc/passive_listener_injectioh.h`. If we take your suggestion to use InternallyRefCounted, I think we expose that type publicly. Should we use a shared_ptr?

Reply:
This took some doing! I was able to make something very much like your design work. The listeners lock is gone, and tcp servers are not created for passive listeners.